reltool: Add load_dot_erlang rel option

okeuday opened the pull request:
This adds a rel tuple to the reltool release specific configuration format
as {rel, Name, Vsn, RelApps, Opts} to support the use of
{rel, Name, Vsn, RelApps, [{load_dot_erlang, false}]} for preventing
the insertion of {apply,{c,erlangrc,[]}} into the release script file
and the release boot file.

This change makes it possible to prevent releases generated with reltool
from attempting to load a ${HOME}/.erlang file which is desirable for
production deployment of Erlang, to have greater control of what BEAM
evaluates.  The ${HOME}/.erlang file lacks visibility and its location on
the filesystem separate from where the release lives means it can be
misused in a way that lacks visibility.

sirihansen commented on lib/reltool/src/reltool_server.erl:
```diff
@@ -1483,6 +1483,18 @@ decode(#sys{rels = Rels} = Sys, [{rel, Name, Vsn, RelApps} | SysKeyVals])
+    Rel1 = lists:foldl(fun(Opt, Rel0) ->
+        case Opt of
+            {load_dot_erlang, Value} when is_boolean(Value) ->
+                Rel0#rel{load_dot_erlang = Value};
```
I think you have forgotten to include reltool.hrl in the commit, since you add this new field in the `rel` record.

sirihansen replied:
I'm sorry, I was looking at an old version of reltool.hrl. Now I see the field already exists.

sirihansen commented:
I think this is a good idea, so please add test and documentation, and add the new `{rel,...}` tuple to the `sys()` type specification in reltool.hrl.

okeuday commented:
@sirihansen I added a test and modified an example in the documentation to show usage.  I also added to the `sys()` type specification.  The commits were squashed into a single commit.  The reltool CT test doesn't get ran by Travis (from what I have seen), so I attempted to run it locally though I got many unrelated failures when using the directions at https://github.com/erlang/otp/wiki/Running-tests#running .  The test case I added works with my local run.

sirihansen commented:
Thanks, it looks good! The only thing I was thinking about is whether or not the `load_dot_erlang` option should be visible in the return of `reltool:get_config/1,3`. It would be somehow incompatible, so should only be done if we find it really useful. Any thoughts about that?

okeuday commented:
@sirihansen I believe that `{rel, Name, Vsn, RelApps, [{load_dot_erlang, false}]}` should be used with all releases to avoid using the `${HOME}/.erlang` file for any release because that file is only a user configuration file and not part of the release (only for experimentation and development, not a dependable release in a specific location).  So, I regard the `load_dot_erlang` option as very useful, but to modify the `reltool:get_config/1,3` return would only make the `rel` tuple a bit more explicit as `{rel, Name, Vsn, RelApps}` becoming `{rel, Name, Vsn, RelApps, [{load_dot_erlang, true}]}` even though that is the same as `{rel, Name, Vsn, RelApps, []}` and `{rel, Name, Vsn, RelApps}`.  Changing the `reltool:get_config/1,3` return doesn't seem appropriate because reltool configuration is normally not provided in an explicit way, it is normally kept minimal and succinct IMO.  Not changing the `reltool:get_config/1,3` return also avoids breaking past usage and the existing tests that expect `{rel, Name, Vsn, RelApps}`.

I understand that not everyone may want to avoid using the `${HOME}/.erlang` file because they may already have releases using it, or they may just feel comfortable using the `${HOME}/.erlang` file for historical reasons.  That is why I assume the default value of `load_dot_erlang` is unable to change and must remain true.  If the default value for `load_dot_erlang` did change, then it would be necessary to change the `reltool:get_config/1,3` return to become explicit, to make the change clearer (i.e., it would intentionally break past use by making the `load_dot_erlang` value visible as the default false, allowing modification to occur if the value wasn't the value desired).

So, while I find the option necessary and extremely useful, I don't think modifying the `reltool:get_config/1,3` return is appropriate because remaining compatible with past usage seems required.

sirihansen commented:
Yes, I think the default value has to remain `true`, and as long as it is not explicitly set, the return value of `get_config/1` should definitely not be any other than `{rel, Name, Vsn, RelApps}`, as it is today. But what if it is set to `false` - then the return of `get_config/1`, if saved and given as input to another reltool instance, would not produce the same release. Is that a problem? (And what if the `InclDefaults` argument to `get_config/3` is `true` - should it be returned then?)

okeuday commented:
@sirihansen I added the additional commit to ensure the `load_dot_erlang` option is present when set to `false`.  That approach has a minimal impact, affecting the default `"no_dot_erlang"` release (as shown in the CT test).  I don't see a problem with that change for compatibility (old config data that happens to have `"no_dot_erlang"` which is consumed will now be correct, so it is an improvement in my view) or style (often people prefer `true`/`false` checked in a `case` though the result has been more minimal with an `if` in the past).  Tell me if I should squash the commits or if there is a problem.

sirihansen commented on lib/reltool/src/reltool_target.erl:
```diff
@@ -201,11 +197,18 @@ do_gen_config(#mod{name = Name,
+              InclDefs) ->
+    RelAppsConfig = [do_gen_config(RA, InclDefs) || RA <- RelApps],
+    if
+        LoadDotErlang =:= true ->
```
To be correct, I think we need to do this only if `InclDefs=:=false`. That is, we should return the options list, not matter the value of load_dot_erlang, if `InclDefs=:=true`. This will of course open for more compatibility issues, but I think we should do it anyway. I'll mark the change as "potentially incompatible" in the release notes.

okeuday replied:
I added the most recent commit to ensure the options list is present with the default when `InclDefs=:=true`.  Does everything looks good?

sirihansen replied:
Yes, it looks good. I'll include it in tests today.

sirihansen commented on lib/reltool/test/reltool_server_SUITE.erl:
```diff
@@ -2107,7 +2134,7 @@ save_config(Config) ->
 		     {boot_rel,"start_clean"},
-                     {rel,"no_dot_erlang","1.0",[]},
+                     {rel,"no_dot_erlang","1.0",[],[{load_dot_erlang,false}]},
 		     {rel,"start_clean","1.0",[]},
```
Failed in tests: reltool_server:save_config also listens to InclDefs, so all 'rel' lines here will show load_dot_erlang.

sirihansen commented on lib/reltool/test/reltool_server_SUITE.erl:
```diff
@@ -2148,7 +2175,7 @@ save_config(Config) ->
 		     {boot_rel,"start_clean"},
-                     {rel,"no_dot_erlang","1.0",[]},
+                     {rel,"no_dot_erlang","1.0",[],[{load_dot_erlang,false}]},
 		     {rel,"start_clean","1.0",[]},
```
InclDefs==true here, so load_dot_erlang is included.

okeuday commented:
@sirihansen Thank you for finding these problems via the internal Ericsson build servers.  However, the current testing documentation is frustrating because it doesn't appear to show how the tests are ran in a way that allows reltool tests to pass.  It seems like the reltool tests require that the `test_server` be installed due to the error typically returned when following the steps at https://github.com/erlang/otp/wiki/Running-tests with the errors like:

    reltool_server_SUITE:create_multiple_standalone failed on line 1135
    Reason: {badmatch,{error,"This system is not installed. The director...}

Do you know what step(s) I am missing from the https://github.com/erlang/otp/wiki/Running-tests documentation (also http://erlang.org/doc/installation_guide/INSTALL.html#how-to-build-and-install-erlang-otp and https://github.com/erlang/otp/blob/master/HOWTO/TESTING.md) that would allow the reltool tests to run?

sirihansen commented:
Great - I'll include the PR in the tests again. And thanks for the updated documentation. I'll suggest an update for `TESTING.md` and `INSTALL.md` as well.
